### PR TITLE
Amrfm 1244 gazebo realsense plugin add setting to specify qo s

### PR DESCRIPTION
**Pull request by ghost**
_For the realsense cameras, in the driver we can specify the QoS to use. We use the SENSOR_DATA QoS there usually. So, to keep our stack compatible, we would like to use the same QoS in simulation as well.
For that, please add a setting "pointcloud_qos" and "color_qos" to the realsense_gazebo_plugin.
In our launchfiles, please set it to sensor_data._

### Setting to specify qos
**file path:**
`/code/ros2_ws/src/navigation2/amr_description/urdf/amr_description.gazebo.xacro`
### option
 ```
 SystemDefaultsQoS
 SensorDataQoS
 ParametersQoS
 ServicesQoS
 ParameterEventsQoS
```
### example:
```
<gazebo>
      <plugin name='${camera_left}' filename='librealsense_gazebo_plugin.so'>
        <prefix>${camera_left}</prefix>
        <depthUpdateRate>${cameras_update_rate}</depthUpdateRate>
        <colorUpdateRate>${cameras_update_rate}</colorUpdateRate>
        <infraredUpdateRate>1.0</infraredUpdateRate>
        <depthTopicName>${camera_left}/aligned_depth_to_color/image_raw</depthTopicName>
        <depthCameraInfoTopicName>${camera_left}/depth/camera_info</depthCameraInfoTopicName>
        <colorTopicName>${camera_left}/color/image_raw</colorTopicName>
        <colorCameraInfoTopicName>${camera_left}/color/camera_info</colorCameraInfoTopicName>
        <infrared1TopicName>${camera_left}/infra1/image_raw</infrared1TopicName>
        <infrared1CameraInfoTopicName>${camera_left}/infra1/camera_info</infrared1CameraInfoTopicName>
        <infrared2TopicName>${camera_left}/infra2/image_raw</infrared2TopicName>
        <infrared2CameraInfoTopicName>${camera_left}/infra2/camera_info</infrared2CameraInfoTopicName>
        <colorOpticalframeName>${camera_left}_color_optical_frame</colorOpticalframeName>
        <depthOpticalframeName>${camera_left}_depth_optical_frame</depthOpticalframeName>
        <infrared1OpticalframeName>${camera_left}_left_ir_optical_frame</infrared1OpticalframeName>
        <infrared2OpticalframeName>${camera_left}_right_ir_optical_frame</infrared2OpticalframeName>
        <rangeMinDepth>0.3</rangeMinDepth>
        <rangeMaxDepth>3.0</rangeMaxDepth>
        <pointCloud>true</pointCloud>
        <pointCloudTopicName>${camera_left}/depth/color/points</pointCloudTopicName>
        <pointCloudCutoff>0.3</pointCloudCutoff>
        <pointCloudCutoffMax>3.0</pointCloudCutoffMax>
        <pointcloud_qos>SystemDefaultsQoS</pointcloud_qos>
        <color_qos>SystemDefaultsQoS</color_qos>
      </plugin>
    </gazebo>
```

### CODE REVIEW

**Review by jplapp — CHANGES_REQUESTED** [review]
logic looks good to me! We want to propose to merge this to the upstream repo later on. For this, we should not change the formatting, e.g. in RealSensePlugin.cpp . Can you align it with the original formatting?

**Comment by ghost** [comment]
> logic looks good to me! We want to propose to merge this to the upstream repo later on. For this, we should not change the formatting, e.g. in RealSensePlugin.cpp . Can you align it with the original formatting?

Ok, will do

File: src/RealSensePlugin.cpp [hunk 1]
```diff
@@ -58,72 +58,75 @@ void RealSensePlugin::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
...
+            pointCloudTopic_ = _sdf->GetValue()->GetAsString();
+        } else if (name == "pointCloudCutoff") {
+            _sdf->GetValue()->Get(pointCloudCutOff_);
+        } else if (name == "pointCloudCutoffMax") {
+            _sdf->GetValue()->Get(pointCloudCutOffMax_);
+        } else if (name == "prefix") {
+            this->prefix = _sdf->GetValue()->GetAsString();
+        } else if (name == "color_qos") {
+            this->colorQos = _sdf->GetValue()->GetAsString(); }
+        else if (name == "pointcloud_qos") {
```
Comment (andriimaistruk) [inline src/RealSensePlugin.cpp]: these are only new, correct?

File: src/ros_qos_specifier.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#include "realsense_gazebo_plugin/ros_qos_specifier.hpp"
+
+image_transport::CameraPublisher ImageTransportWithSpecifiedQos::advertise_camera(
+        const std::string & base_topic, uint32_t queue_size, rmw_qos_profile_t qos_profile) {
+    rmw_qos_profile_t custom_qos = qos_profile;
+    custom_qos.depth = queue_size;
+    return image_transport::CameraPublisher(ad_node_p.get(), base_topic, custom_qos);
```
Comment (andriimaistruk) [inline src/ros_qos_specifier.cpp:7]: what's the point of this return value? who receives it?

File: src/ros_qos_specifier.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#include "realsense_gazebo_plugin/ros_qos_specifier.hpp"
+
+image_transport::CameraPublisher ImageTransportWithSpecifiedQos::advertise_camera(
+        const std::string & base_topic, uint32_t queue_size, rmw_qos_profile_t qos_profile) {
+    rmw_qos_profile_t custom_qos = qos_profile;
```
Comment (andriimaistruk) [inline src/ros_qos_specifier.cpp]: no need for intermediate copy, you already have done copying in the argument list

File: src/gazebo_ros_realsense.cpp [hunk 4]
```diff
@@ -49,21 +49,36 @@ void GazeboRosRealsense::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
...
-  this->depth_pub_ = this->itnode_->advertiseCamera(
-    cameraParamsMap_[DEPTH_CAMERA_NAME].topic_name, 2);
-
-  if (pointCloud_) {
-    this->pointcloud_pub_ = this->node_->create_publisher<sensor_msgs::msg::PointCloud2>(
-      pointCloudTopic_, rclcpp::SystemDefaultsQoS());
-  }
+    this->itnode_.reset(new ImageTransportWithSpecifiedQos(this->node_));
+
+    itnode_->specify_color_qos(this->color_pub_,cameraParamsMap_[COLOR_CAMERA_NAME].topic_name,colorQos);
```
Comment (andriimaistruk) [inline src/gazebo_ros_realsense.cpp]: `this->` hint is reduntant, compiler can find its attributes within the object

File: src/RealSensePlugin.cpp (hunk 1 above)
Comment (ghost) [reply src/RealSensePlugin.cpp]: yes, it is

File: src/ros_qos_specifier.cpp (hunk 3 above)
Comment (ghost) [reply src/ros_qos_specifier.cpp]: fixed

File: src/gazebo_ros_realsense.cpp (hunk 4 above)
Comment (ghost) [reply src/gazebo_ros_realsense.cpp]: ok

**Review by jplapp — CHANGES_REQUESTED** [review]
there are still many formatting differences between the files. You can open the "files" tab to see them. You can use some diff tool such as the one in pycharm / Clion / vscode to compare this branch to master, and revert the "hunks" which are only formatting changes.

File: src/gazebo_ros_realsense.cpp [hunk 5]
```diff
@@ -49,22 +49,36 @@ void GazeboRosRealsense::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
...
+            RCLCPP_INFO(node_->get_logger(), "Gazebo ROS Realsense plugin -> publisher created using ServicesQoS.");
+        } else if(pointCloudQos=="ParameterEventsQoS") {
+            pointcloud_pub_ = node_->create_publisher<sensor_msgs::msg::PointCloud2>(
+                    pointCloudTopic_, rclcpp::ParameterEventsQoS());
+            RCLCPP_INFO(node_->get_logger(), "Gazebo ROS Realsense plugin -> publisher created using ParameterEventsQoS.");
+        } else {
+            pointcloud_pub_ = node_->create_publisher<sensor_msgs::msg::PointCloud2>(
+                    pointCloudTopic_, rclcpp::SystemDefaultsQoS());
+            RCLCPP_INFO(node_->get_logger(), "Gazebo ROS Realsense plugin -> publisher created using SystemDefaultsQoS.");
+        }
```
Comment (ghost) [inline src/gazebo_ros_realsense.cpp]: to reduce code you could just have a variable for the QOS in the if else statements and at the end create the publisher

File: include/realsense_gazebo_plugin/RealSensePlugin.h [hunk 6]
```diff
@@ -126,6 +126,8 @@ class RealSensePlugin : public ModelPlugin {
 
   bool pointCloud_ = false;
   std::string pointCloudTopic_;
+  std::basic_string<char> pointCloudQos;
+  std::basic_string<char>colorQos;
```
Comment (ghost) [inline include/realsense_gazebo_plugin/RealSensePlugin.h]: space?

File: src/ros_qos_specifier.cpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    RCLCPP_INFO(this->ad_node_p->get_logger(), "Gazebo ROS Realsense plugin -> advertise camera using ServicesQoS.");
+  } else if(colorQos=="ParameterEventsQoS") {
+    cam_color_pub = this->advertise_camera(
+            topic_name, 2,rclcpp::ParameterEventsQoS().get_rmw_qos_profile());
+    RCLCPP_INFO(this->ad_node_p->get_logger(), "Gazebo ROS Realsense plugin -> advertise camera using ParameterEventsQoS.");
+  } else  {
+    cam_color_pub = this->advertise_camera(
+            topic_name, 2,rclcpp::SystemDefaultsQoS().get_rmw_qos_profile());
+    RCLCPP_INFO(this->ad_node_p->get_logger(), "Gazebo ROS Realsense plugin -> advertise camera using SystemDefaultsQoS.");
+  }
```
Comment (ghost) [inline src/ros_qos_specifier.cpp]: same comment as above, could be made simpler imo

**Comment by ghost** [comment]
In general I don't see the point of going through the trouble of parametrizing the QOS. Why don't we just set the QOS we need until this is solved https://github.com/ros-perception/image_common/issues/156?

**Comment by OleksandrMoskaliuk8800** [comment]
> In general I don't see the point of going through the trouble of parametrizing the QOS. Why don't we just set the QOS we need until this is solved [ros-perception/image_common#156](https://github.com/ros-perception/image_common/issues/156)?

In image transporter class we don't have possibility to do that, only change it for all image, because there no functionality to make that. For that purpose I use inheritance and added new method to specify qos

File: include/realsense_gazebo_plugin/RealSensePlugin.h (hunk 6 above)
Comment (OleksandrMoskaliuk8800) [reply include/realsense_gazebo_plugin/RealSensePlugin.h:130]: yes, just string type

Comment (ghost) [reply include/realsense_gazebo_plugin/RealSensePlugin.h]: fixed

File: src/gazebo_ros_realsense.cpp (hunk 5 above)
Comment (ghost) [reply src/gazebo_ros_realsense.cpp]: ok, will do

File: src/ros_qos_specifier.cpp (hunk 7 above)
Comment (ghost) [reply src/ros_qos_specifier.cpp]: ok

File: src/gazebo_ros_realsense.cpp [hunk 8]
```diff
@@ -49,22 +49,32 @@ void GazeboRosRealsense::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
   this->camera_info_manager_.reset(
     new camera_info_manager::CameraInfoManager(this->node_.get(), this->GetHandle()));
 
-  this->itnode_.reset(new image_transport::ImageTransport(this->node_));
+  itnode_.reset(new ImageTransportWithSpecifiedQos(node_));
```
Comment (jplapp) [inline src/gazebo_ros_realsense.cpp:52]: can `this` really be removed? does it still work then?

**Review by jplapp — COMMENTED** [review]
looks good in the end to me.
@tnajjar the realsense driver for actual realsense devices offers the same options. In general, it's useful to use SensorDataQoS here, that should be better for performance than RELIABLE transport.

File: src/gazebo_ros_realsense.cpp (hunk 8 above)
Comment (OleksandrMoskaliuk8800) [reply src/gazebo_ros_realsense.cpp:52]: Yes, it can works be without "this->", compiler can find its attributes within the object

Comment (ghost) [reply src/gazebo_ros_realsense.cpp:52]: Yes, it can works without "this->", compiler can find its attributes within the object

**Review by ghost — APPROVED** [review]
thanks for the clarifications

**Review by jplapp — APPROVED** [review]
thanks for the clarification

**Comment by tonynajjar** [comment]
@OleksandrMoskaliuk8800 I got pinged on https://github.com/ros-perception/image_common/issues/156#issuecomment-1013601024. You know more than me on this topic, can you check if the comment would be useful for us?

@jplapp kind reminder that our fork is public, i guess it's fine

**Pull request merged**